reframe/frontend/executors/policies.py

```diff
@@ -307,13 +394,15 @@ def exit(self):
         pollrate = PollRateFunction(0.2, 60)
         num_polls = 0
         t_start = datetime.now()
-        while self._running_tasks or self._retired_tasks:
+        while (self._running_tasks or self._waiting_tasks):
```

Comment:
You will miss tasks here with this condition. As soon as you schedule your last batch of tasks, you won't be able to finalize them, because you will be out of the loop. That's why I was checking also the `retired_tasks` list. You should check that all the intermediate lists have be flushed.

Reply:
A task is removed from `running_tasks` inside `on_task_exit` or inside `on_task_failure`. In the first case it is called by `_poll_tasks()` so in the next loop. In the second case the failed task will not have to be cleaned up, as well as the tasks the failed task was depending on. So which case would be caught if I was checking for `completed_tasks` or `ready_tasks`? Not all completed tasks should be cleaned up in the end.

Reply:
I was thinking mostly about the case of success rather that failure, but your explanation makes sense.